sql/row: fix multi-range reads from PCR standby

michae2 opened the pull request:
Reads from tables with external row data (i.e. reads from a PCR standby cluster) need to use the fixed timestamp specified by the external row data. This timestamp might be different from the transaction timestamp, so we were explicitly setting BatchRequest.Timestamp in kv_batch_fetcher.

The KV API only allows BatchRequest.Timestamp to be set for non-transactional requests (i.e. requests sent with a NonTransactionalSender, which is a CrossRangeTxnWrapperSender in this case). We were using a NonTransactionalSender, but this had two problems:

1. CrossRangeTxnWrapperSender in turn sends the BatchRequest with a transactional sender, which again does not allow BatchRequest.Timestamp to be set.
2. CrossRangeTxnWrapperSender uses `kv.(*Txn).CommitInBatch`, which does not provide the 1-to-1 request-response guarantee required by txnKVFetcher. It is `kv.(*Txn).Send` which provides this guarantee.

Because of these two problems, whenever the txnKVFetcher would send a multi-range-spanning BatchRequest to CrossRangeTxnWrapperSender, it would either fail with a "transactional request must not set batch timestamp" error or would return an unexpected number of responses, violating the txnKVFetcher's assumed mapping from request to response.

To fix both these problems, instead of using a NonTransactionalSender, change the txnKVFetcher to open a new root transaction with the correct fixed timestamp, and then use txn.Send.

Fixes: #132608

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132854)
<!-- Reviewable:end -->

michae2 committed:
sql/row: fix multi-range reads from PCR standby

Reads from tables with external row data (i.e. reads from a PCR standby
cluster) need to use the fixed timestamp specified by the external row
data. This timestamp might be different from the transaction timestamp,
so we were explicitly setting BatchRequest.Timestamp in
kv_batch_fetcher.

The KV API only allows BatchRequest.Timestamp to be set for
non-transactional requests (i.e. requests sent with a
NonTransactionalSender, which is a CrossRangeTxnWrapperSender in this
case). We were using a NonTransactionalSender, but this had two
problems:

1. CrossRangeTxnWrapperSender in turn sends the BatchRequest with a
   transactional sender, which again does not allow
   BatchRequest.Timestamp to be set.
2. CrossRangeTxnWrapperSender uses `kv.(*Txn).CommitInBatch`, which does
   not provide the 1-to-1 request-response guarantee required by
   txnKVFetcher. It is `kv.(*Txn).Send` which provides this guarantee.

Because of these two problems, whenever the txnKVFetcher would send a
multi-range-spanning BatchRequest to CrossRangeTxnWrapperSender, it
would either fail with a "transactional request must not set batch
timestamp" error or would return an unexpected number of responses,
violating the txnKVFetcher's assumed mapping from request to response.

To fix both these problems, instead of using a NonTransactionalSender,
change the txnKVFetcher to open a new root transaction with the correct
fixed timestamp, and then use txn.Send.

Fixes: #132608

Release note: None

DrewKimball reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> assuming nested RootTxns are allowed,. I'll defer to @nvanbenschoten for final approval.

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132854#-:-O9S3Tp26bT_dZYSEIO3:bg6itxr)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt, @mgartner, @michae2, and @nvanbenschoten)

___
*[`pkg/sql/row/kv_batch_fetcher.go` line 318 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132854#-O9S2k82EfFqrIUjOvz9:-O9S2k82EfFqrIUjOvzA:bqn6s93) ([raw file](https://github.com/cockroachdb/cockroach/blob/d923e68a1739895d197be2425900796106d59153/pkg/sql/row/kv_batch_fetcher.go#L318)):*
> ```Go
> 		// required by txnKVFetcher.
> 		var res *kvpb.BatchResponse
> 		err := db.TxnWithAdmissionControl(
> ```

[nit] It's probably fine since this is a per-batch operation, but do you think there would be any benefit to opening this nested transaction once, and then referencing it from this sendFunc?
___
*[`pkg/ccl/crosscluster/physical/standby_read_test.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132854#-O9S3E-kBWE-n-xmeRX6:-O9S3E-kBWE-n-xmeRX7:b7hp2ni) ([raw file](https://github.com/cockroachdb/cockroach/blob/d923e68a1739895d197be2425900796106d59153/pkg/ccl/crosscluster/physical/standby_read_test.go#L38)):*
> ```Go
> 		{stmt: `CREATE TABLE abc (a INT PRIMARY KEY, b INT, c JSONB)`},
> 		{stmt: `ALTER TABLE abc CONFIGURE ZONE USING range_max_bytes = 64 << 20, range_min_bytes = 1 << 20`},
> 		// Insert 2m rows, which should be split into two ranges.
> ```

[nit] Rather than setting the max_bytes, would it be more reliable / faster to explicitly set range boundaries, and insert less rows?


<!-- Sent from Reviewable.io -->

dt commented on pkg/ccl/crosscluster/physical/standby_read_test.go:
```diff
@@ -0,0 +1,105 @@
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	skip.UnderDuress(t, "slow test")
```
Unless you specifically are wanting to observe the behavior of the PCR job -- that is, copying bytes from one cluster to another and publishing progress updates reflecting what is copied -- I probably wouldn't use a PCR job in this test since that means you have replicate millions of rows and sit around in the associated slow WaitUntilReplicatedTime calls.

As far as I can tell, what you are intending to test here is the behavior of external-row-data query execution that fetches from a second tenant tenant, right? I'd probably suggest a simpler, faster test that just runs one cluster that has two tenants, one that you write to and one that reads from that first one as external data:

Concretely, in a test (somewhere in the sql tree?) I'd:
* create tenant twice
* write to the first tenant
* call `sql/catalog/replication.SetupOrAdvanceStandbyReaderCatalog` on the second (passing now())
* run your query on conn to the second

This way there is no PCR job, and you'll just directly move the read-as-of ts up after each write you want to observe, there's no waiting for a job to notice the change, copy it, then update progress.

michae2 replied:
Good point, I've rewritten the test to use two tenants in a TestCluster. It now takes about 12s instead of > 60.

dt replied:
Nice! Looks great. Foundations might have some knobs that speed up the lease waits to get cut that down even more.

michae2 reviewed (commented):
> assuming nested RootTxns are allowed

AIUI this transaction isn't really _nested_ (there's nothing connecting it with the "outer" transaction), rather, it is simply a separate root transaction. This is also what [CrossRangeTxnWrapperSender](https://github.com/cockroachdb/cockroach/blob/7d5fd97b51449704187d6e796cc6d35dce432b4a/pkg/kv/db.go#L229-L246) does under the hood, so it seemed safe enough.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132854#-:-O9lEGIC7FbmMj4HJ6bo:blrlafj)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, and @nvanbenschoten)

___
*[`pkg/sql/row/kv_batch_fetcher.go` line 318 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132854#-O9S2k82EfFqrIUjOvz9:-O9lFRXP7eOiwsGZNn5B:bg3m1yb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d923e68a1739895d197be2425900796106d59153/pkg/sql/row/kv_batch_fetcher.go#L318)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] It's probably fine since this is a per-batch operation, but do you think there would be any benefit to opening this nested transaction once, and then referencing it from this sendFunc?
</blockquote></details>

Good point, there might be a performance benefit. I've left it as a TODO for now.
___
*[`pkg/ccl/crosscluster/physical/standby_read_test.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132854#-O9S3E-kBWE-n-xmeRX6:-O9lFaDpEF68PiN-Fj3C:boy9h3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/d923e68a1739895d197be2425900796106d59153/pkg/ccl/crosscluster/physical/standby_read_test.go#L38)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Rather than setting the max_bytes, would it be more reliable / faster to explicitly set range boundaries, and insert less rows?
</blockquote></details>

Yes, you're right. This wasn't possible when I was using PCR, but now it is. Done.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> assuming nested RootTxns are allowed

They are allowed, but we do need to be careful about deadlocks if the outer txn has acquired a lock that the inner txn might decide to block on. Or if the outer txn has acquired a lock that some other txn might decide to block on, which itself has acquired locks. If we are going to do this, we need to make sure that the inner txn doesn't block directly or indirectly on the outer one.

> Reviewed 2 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132854#-:-O9lJ6uf2H__RbWTqA3F:b-juj75w)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @dt, and @mgartner)



<!-- Sent from Reviewable.io -->

dt commented:
The “inner” txn here is reading from a wholly different tenant (and doing so at a fixed ts) so I would be quite surprised if the “outer” txn ever held a lock on which it blocked.

DrewKimball reviewed (approved):
> Reviewed 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132854)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @michae2)



<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
TFTRs!

bors r=DrewKimball,nvanbenschoten,dt

> They are allowed, but we do need to be careful about deadlocks if the outer txn has acquired a lock that the inner txn might decide to block on. Or if the outer txn has acquired a lock that some other txn might decide to block on, which itself has acquired locks. If we are going to do this, we need to make sure that the inner txn doesn't block directly or indirectly on the outer one.

These are all read-only transactions, so neither inner nor outer txn can acquire locks.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132854#-:-O9pAr2s5yuiwHYcc5KY:bwuhxm9)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mgartner)



<!-- Sent from Reviewable.io -->
